script: Fix scrollIntoView positioning

abdelrahman1234567 opened the pull request:
Previously, scrollIntoView would incorrectly scroll all ancestor elements, leading to unexpected behavior when elements had complex positioning contexts.

Changes:
- Get correct positioning for scrolling container using HTMLElement offset properties
- Correctly calculate element position in scroller's content coordinate system. We need to determine if the element's coordinates need to be adjusted relative to the scrolling container

These changes are more accurate afterwards when smooth scrolling is going to be introduced to Servo.

This PR is based on #38773 for updating test expectations that are based on both PRs together, so please review #38773 first.

Testing: Existing WPT tests

Commit:
script: implementing scrollend event

Implementing and handling scrollend even based on spec: https://drafts.csswg.org/cssom-view/#eventdef-document-scrollend

Signed-off-by: abdelrahman1234567 <abdelrahman.hossameldin.awadalla@huawei.com>

mrobinson commented on components/script/dom/element.rs:
```diff
@@ -998,14 +1001,71 @@ impl Element {
+    /// Check if an element is in the containing block chain for scrollIntoView
+    /// Only elements in the containing block chain should be scrolled by scrollIntoView
+    fn is_in_containing_block_chain(&self, ancestor: &Node) -> bool {
+        if ancestor.is::<Document>() {
+            true // Document viewport is always in the containing block chain
+        } else if ancestor.is::<Element>() {
```
Please use an early return here.

mrobinson commented on components/script/dom/element.rs:
```diff
@@ -998,14 +1001,71 @@ impl Element {
+            }
+
+            false
+        } else {
+            false
+        }
+    }
```
Please use an early return at the top of the function instead of the large nested block.

mrobinson commented on components/script/dom/element.rs:
```diff
@@ -998,14 +1001,71 @@ impl Element {
+                    if let Some(style) = element.style() {
+                        let position = style.get_box().clone_position();
+
+                        // If this element is absolutely or fixed positioned,
+                        // it establishes a new containing block - stop here
+                        if position.is_absolutely_positioned() && !std::ptr::eq(element, self) {
+                            return false;
+                        }
```
Hrm. This is a bit odd. What element forms a containing block is determined by both the `position` attribute of the element in question and various aspects of the ancestor in question (https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_display/Containing_block).

The wording here is quite odd. If an element is `position: static` *all* ancestors establish a containing block.

I wonder a bit if we should just ask layout here to answer this query. At the very least we could reuse `StyleExt` to figure out which elements actually form a containing block for `self`.

abdelrahman1234567 replied:
So what I did is, for each time we will need to scroll into view i will send a list of all ancestors that establish scrolling_box to query on them to layout which one of them has element in their containing block and return all ancestors that does have that to apply on them the scrolling.

mrobinson commented on components/script/dom/element.rs:
```diff
@@ -998,14 +1001,71 @@ impl Element {
     ) -> ScrollPosition {
-        let target_bounding_box = self.upcast::<Node>().content_box().unwrap_or_default();
+        // Get the target element's stable position using offset properties
+        let target_bounding_box = if let Some(html_element) = self.downcast::<HTMLElement>() {
```
Please use a `map()` here with `unwrap_or_else` below.

mrobinson commented on components/script/dom/element.rs:
```diff
@@ -1044,150 +1104,144 @@ impl Element {
-                    }
+            // Calculate element position in scroller's content coordinate system
+            // We need to determine if the element's coordinates need to be adjusted relative to the scrolling container
+            let target_element = if let Some(html_element) = self.downcast::<HTMLElement>() {
```
Please use `unwrap_or_else` here.

abdelrahman1234567 replied:
I think it is better to do map_or_else

mrobinson commented on components/script/dom/window.rs:
```diff
@@ -1550,8 +1550,15 @@ impl WindowMethods<crate::DomTypeHolder> for Window {
         let y = options.top.unwrap_or(0.0f64);
-        self.ScrollBy_(x, y);
-        self.scroll(x, y, options.parent.behavior);
+
+        let scroll_offset = self.scroll_offset();
+        // Step 3
+        let left = x + scroll_offset.x as f64;
+        // Step 4
+        let top = y + scroll_offset.y as f64;
+
+        // Step 5
+        self.scroll(left, top, options.parent.behavior);
```
Please fill in the step text from the specification.

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+        if let Some(element) = ancestor_node.as_element() {
+            let style_data = element.style_data();
+            let position = style_data.styles.primary().get_box().position;
+            if position == Position::Fixed {
```
I guess this should ignore elements with `display: contents`

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+            }
+        }
+    }
+    let is_inside_fixed_context = target_position == Position::Fixed || found_fixed;
```
So the loop above can be avoided if `target_position == Position::Fixed`?
I don't really follow the logic, though. Being fixed and being inside a fixed seem completely different situations.

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+        // fixed element must not scroll its containing frame; however, it
+        // must scroll further ancestor scrollers as the element isn't fixed
+        // in relation to them.
+        containing_block_chain.insert(target_node.opaque().0, true);
```
Why are you adding the target in the containing block chain?

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+
+                // Check if this ancestor establishes a containing block for fixed descendants
+                // properties like transform, will-change, contain, etc. can establish this
+                if style.establishes_containing_block_for_all_descendants(FragmentFlags::empty()) {
```
You can't just use `FragmentFlags::empty()` here.

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+                    continue;
+                }
+
+                containing_block_chain.insert(ancestor_node.opaque().0, true);
```
So if there is no containing block, you insert all ancestors as containing blocks? I don't follow.

abdelrahman1234567 replied:
I believe if there is no containing block then we should insert the initial C.B and break the loop. It is modified now

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+                        // any positioned element (position != static) or elements with
+                        // transform, will-change, etc. establish containing blocks
+                        if style.establishes_containing_block_for_absolute_descendants(
+                            FragmentFlags::empty(),
```
Ditto `FragmentFlags::empty()`

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+                if !found_absolute_cb {
+                    // Include all ancestors up to the initial containing block
+                    for &ancestor_node in ancestor_nodes {
+                        containing_block_chain.insert(ancestor_node.opaque().0, true);
```
Again, I don't understand this logic

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+                        let overflow_x = style.get_box().overflow_x;
+                        let overflow_y = style.get_box().overflow_y;
+
+                        // An element establishes a containing block for static descendants if:
+                        // - It has position != static
+                        // - Or it has overflow != visible (establishes BFC)
```
That's not true.

If I'm not missing anything, any ancestor that generates a box which isn't either a non-replaced inline or an anonymous block should establish a containing block.

A BFC is a completely unrelated concept, and it's not just established by `overflow`.

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+            Position::Relative => {
+                containing_block_chain.insert(target_node.opaque().0, true);
+
+                // For relative elements, the containing block is the same as for static
```
Yes, it's the same, so merge the cases instead of copy-pasting.

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+            Position::Sticky => {
+                containing_block_chain.insert(target_node.opaque().0, true);
+
+                // For sticky elements, the containing block is established by the nearest ancestor
+                // that is a scroll container or the viewport
```
Not very familiar with sticky, but I'm quite sure this should be the same as for static and relative.

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+) -> Vec<usize> {
+    // A sequence of successive containing blocks that form an
+    // ancestor-descendant chain through the containing block relation.
+    let mut containing_block_chain = HashMap::new();
```
Why HashMap if you are always setting the value to true? Use HashSet instead?

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,215 @@ fn transform_au_rectangle(
+                unreachable!("Fixed position should be handled above");
+            },
+        }
+    }
```
At this point, if you found the containing block, you inserted it into `containing_block_chain` and then broke the loop. So `containing_block_chain` doesn't have the chain, jut the first one.

abdelrahman1234567 replied:
You are right, I modified the whole implementation now. It was just a draft before because i was trying several things. 

Now, I check the chain by jumping from one ancestor to the next ancestor in the containing block chain based on the rules in CSS spec.

abdelrahman1234567 commented:
@Loirooriol Hey thanks for your review. I am still trying to figure out the containing block chain and the rules that sets each element and its containing blocks and which should be scrolled and which shouldn't. The changes here is still a draft because I was working on tracing the wpt tests to get more examples and corner cases. Since that this implementation is still a draft and still failing most of wpt tests, i will keep working on it and and will let you know once it is done

Loirooriol commented on components/layout/query.rs:
```diff
@@ -1182,3 +1184,126 @@ fn transform_au_rectangle(
+                    if let Some(element) = ancestor_node.as_element() {
+                        let style_data = element.style_data();
+                        let style = style_data.styles.primary();
+                        // An element establishes a containing block for static descendants if:
+                        // - It is a block container or establishes a formatting context
+                        let display = style.get_box().display;
+                        if display != Display::None && display != Display::Contents {
```
You need to check that the element doesn't generate a non-replaced inline box.

Commit:
script: Fix scrollIntoView positioning and checking containing block chain

Previously, scrollIntoView would incorrectly scroll all ancestor elements, leading to unexpected behavior when elements had complex positioning contexts. This change implements proper containing block chain validation as per the CSS specification.

Changes:

Querying on ancestors having element in its containing block to scroll them too
Get correct positioning for scrolling container using HTMLElement offset properties
Correctly calculate element position in scroller's content coordinate system. We need to determine if the element's coordinates need to be adjusted relative to the scrolling container
This ensures scrollIntoView only affects the minimal set of scrolling containers necessary to bring the element into view. In addition, these changes are more accurate afterwards when smooth scrolling is going to be introduced to Servo!

Testing: Existing WPT tests

Signed-off-by: abdelrahman1234567 <abdelrahman.hossameldin.awadalla@huawei.com>

mrobinson reviewed (commented):
A few minor comments, but seems like an improvement...
